### PR TITLE
[ENH] Added Spatio-Temporal DBSCAN

### LINKED ISSUE #7054: Spatio-Temporal Clustering
Hey there,

I am working with collective behavioral data. My goal is to cluster agents in space (ST) and time, so as input I have the coordinates of the agents at each time point and as output I want to have the label for each agent and each timestamp of the assignment to the cluster. See the figure for a graphical representation [adopted from 1]:

image

[Here is the repository](https://github.com/eren-ck/spatio-temporal-clustering-benchmark) for the paper that implements most of the existing algorithms [2]. [Here is the dataset](https://zenodo.org/records/5495226) for the paper.

**Is your feature request related to a problem? Please describe.**
Implementation of the ST-clustering models

**Describe the solution you'd like**
Sktime interface for ST-clustering

**Additional context**
[This](https://github.com/GISerWang/Spatio-temporal-Clustering) github page may also be relevant. 

**References**

1. Ansari, M. Y., Ahmad, A., Khan, S. S., Bhushan, G., & Mainuddin. (2020). Spatiotemporal clustering: A review. Artificial Intelligence Review, 53(4), 2381–2423. https://doi.org/10.1007/s10462-019-09736-1
2. Cakmak, E., Plank, M., Calovi, D. S., Jordan, A., & Keim, D. (2021). Spatio-temporal clustering benchmark for collective animal behavior. Proceedings of the 1st ACM SIGSPATIAL International Workshop on Animal Movement Ecology and Human Mobility, 5–8. https://doi.org/10.1145/3486637.3489487

### PR DESCRIPTION
**Pull request by VectorNd**
#### Reference Issues/PRs
Fixes #7054 

#### What does this implement/fix? Explain your changes.
Implemented support for spatio-Temporal DBSCAN algorithm 

#### Does your contribution introduce a new dependency? If yes, which one?
No

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
+"""
+ST-DBSCAN - fast scalable implementation of ST DBSCAN
+            scales also to memory by splitting into frames
+            and merging the clusters together
+"""
+
+__author__ = ["VectorNd"]
```
Comment (vagechirkov) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:7]:
```suggestion
__author__ = ["eren-ck", "VectorNd"]
```

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+from sklearn.cluster import DBSCAN
+from sklearn.utils import check_array
+from sklearn.neighbors import NearestNeighbors
+
+import warnings
+
+class ST_DBSCAN(BaseClusterer):
+    """
+    A class to perform the ST_DBSCAN clustering
+    Parameters
```
Comment (fkiraly) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:25]: newlines are missing in the docstrings, e.g., before section headers - ensure to conform with the code formatting requirements. Here's how you can set them up on your computer, automatically: https://www.sktime.net/en/latest/developer_guide/coding_standards.html

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+from scipy.spatial.distance import pdist, squareform
+from scipy.sparse import coo_matrix
+
+from sklearn.cluster import DBSCAN
+from sklearn.utils import check_array
+from sklearn.neighbors import NearestNeighbors
+
+import warnings
+
+class ST_DBSCAN(BaseClusterer):
```
Comment (fkiraly) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:22]: classes should be `UpperCamelCase`, i.e., no underscores.

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        """
+
+        X = check_array(X)
+
+        if not self.eps1 > 0.0 or not self.eps2 > 0.0 or not self.min_samples > 0.0:
+            raise ValueError('eps1, eps2, minPts must be positive')
+
+        n, m = X.shape
+
+        if len(X) < 20000:
```
Comment (fkiraly) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:93]: this seems like an arbitrary distinction. I would expose the choice which version of the algorithm to run as a hyper-parameter.

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+            treated as spatial coordinates. The structure should look like this [[time_step1, x, y], [time_step2, x, y]..]
+            For example 2D dataset:
+            array([[0,0.45,0.43],
+            [0,0.54,0.34],...])
+        Returns
+        -------
+        self
+        """
+
+        X = check_array(X)
```
Comment (fkiraly) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:86]: check_array is not needed in sktime, the boilerplate will do checking and conversion already.

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        State required:
+            Requires state to be "fitted".
+
+        Returns
+        -------
+        fitted_params : dict with str keys
+            fitted parameters, keyed by names of fitted parameter
+        """
+
+        fitted_params = {
```
Comment (fkiraly) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:254]: not all of these are fitted parameters, some are hyper-parameters from the init.

I would suggest, just delete `_get_fitted_params`, it automatically exposes any attribute ending in underscord.

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+        Returns
+        -------
+        params : dict or list of dict, default = {}
+            Parameters to create testing instances of the class
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`
+        """
+
```
Comment (fkiraly) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:292]: I would also suggest to add a minimal, e.g., empty parameter set, to test the defaults.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice!

Some comments above, I think they are minor but important as they are about API compliance, and the interface.

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        'braycurtis', 'canberra', 'chebyshev', 'cityblock', 'correlation',
+        'cosine', 'dice', 'euclidean', 'hamming', 'jaccard', 'jensenshannon',
+        'kulsinski','manhattan'.
+    n_jobs : int or None, default=-1
+        The number of processes to start -1 means use all processors 
+    Attributes
+    ----------
+    labels : array, shape = [n_samples]
+        Cluster labels for the data - noise is defined as -1
+    """
```
Comment (fkiraly) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:46]: The docstring should reference the original implementation and/or the authors or papers, where relevant

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        The used distance metric - more options are
+        'braycurtis', 'canberra', 'chebyshev', 'cityblock', 'correlation',
+        'cosine', 'dice', 'euclidean', 'hamming', 'jaccard', 'jensenshannon',
+        'kulsinski','manhattan'.
+    n_jobs : int or None, default=-1
+        The number of processes to start -1 means use all processors 
+    Attributes
+    ----------
+    labels : array, shape = [n_samples]
+        Cluster labels for the data - noise is defined as -1
```
Comment (vagechirkov) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:45]:
```suggestion
        Cluster labels for the data - noise is defined as -1
        
    References
    ----------
    .. [1] D. Birant and A. Kut, “ST-DBSCAN: An algorithm for clustering spatial-temporal data,” Data Knowl. Eng., vol. 60, no. 1, pp. 208–221, Jan. 2007, doi: [10.1016/j.datak.2006.01.013](https://doi.org/10.1016/j.datak.2006.01.013).
    .. [2] I. Peca, G. Fuchs, K. Vrotsou, N. Andrienko, and G. Andrienko, “Scalable Cluster Analysis of Spatial Events,” 2012, The Eurographics Association. doi: [10.2312/PE/EUROVAST/EUROVA12/019-023](https://doi.org/10.2312/PE/EUROVAST/EUROVA12/019-023).
    ...[3] E. Cakmak, M. Plank, D. S. Calovi, A. Jordan, and D. Keim, “Spatio-temporal clustering benchmark for collective animal behavior,” in Proceedings of the 1st ACM SIGSPATIAL International Workshop on Animal Movement Ecology and Human Mobility, Beijing China: ACM, Nov. 2021, pp. 5–8. doi: [10.1145/3486637.3489487](https://doi.org/10.1145/3486637.3489487).
```

**Comment by vagechirkov** [comment]
Hey, thanks @VectorNd for the implementation! Happy to work on this PR (as far as I understand, you would need to allow committing changes) 🙂

File: sktime/clustering/spatio_temporal/_st_dbscan.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+
+                row = time_sp.nonzero()[0]
+                column = time_sp.nonzero()[1]
+                v = np.array(euc_sp[row, column])[0]
+
+                dist_sp = coo_matrix((v, (row, column)), shape=(n, n))
+                dist_sp = dist_sp.tocsc()
+                dist_sp.eliminate_zeros()
+
+                db = DBSCAN(eps=self.eps1,
```
Comment (vagechirkov) [inline sktime/clustering/spatio_temporal/_st_dbscan.py:130]: I believe there may be other clustering algorithms here (e.g. `OPTICS`). I wonder if the spatial clustering method could be passed as a parameter 🤔 Otherwise `st_optict`, `st_hdbscan`, etc. will have a lot of duplicated code

Comment (VectorNd) [reply sktime/clustering/spatio_temporal/_st_dbscan.py:130]: Yeah , we can do this . I have only added code for one algo others are remaining

**Review by VectorNd — COMMENTED** [review]
Thanks for the review

File: sktime/clustering/spatio_temporal/_st_dbscan.py (hunk 9 above)
Comment (fkiraly) [reply sktime/clustering/spatio_temporal/_st_dbscan.py:45]: @vagechirkov, is this all the same algorithm? Maybe a more specific pointer (section etc) would be appropriate, if some of these are review papers. Also, there seem to be an extra dot in front of the 3.

File: sktime/clustering/spatio_temporal/_st_dbscan.py (hunk 10 above)
Comment (fkiraly) [reply sktime/clustering/spatio_temporal/_st_dbscan.py:130]: excellent thought! Although this would make most parameteres obsolete, in similar cases where there was a very popular choice and a modular option, we added two estimators, one with hard-coded component, one with a modular component.

File: sktime/clustering/spatio_temporal/_st_dbscan.py (hunk 9 above)
Comment (vagechirkov) [reply sktime/clustering/spatio_temporal/_st_dbscan.py:45]: The first paper presents the algorithm, the second paper is the extension that describes the splitting of the data into frames (it is more relevant for the `_fit_frame_split` method). The third paper presents the benchmark and the code, which is almost identical to the current implementation. I guess I can point to the methods and results section in the third paper, but the paper is 4 pages long 🤷‍♂️

**Comment by fkiraly** [comment]
FYI, code formatting still seems to be failing - see https://www.sktime.net/en/latest/developer_guide/coding_standards.html, and please let us know on dev-chat if you are facing issues with setting this up.

**Comment by vagechirkov** [comment]
Hey @VectorNd, I assumed you were not planning to continue working on this issue, so I opened another PR with the same content. Please let me know if this is not the case and you still intended to work on this issue.  

@fkiraly I have tried to incorporate all the comments from this PR into the new one #7353.
